keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,297 @@
+    * launches new **migration worker** `job` to migrate the resource type. The
+      `job` is annotated with the generation.
+    * makes sure there is only one **migration worker** `job` running at a time,
+      to avoid overloading the apiserver.
+    * updates `migration.status.observedGeneration`.
+* watches the **migration worker** jobs. Updates
+  `migration.status.completedGeneration` when a job completes.
+* short-circuit: if `migration.spec.groupVersions` shows that the resource has
+  only one version, then its `.status.observedGeneration` and
+  `.status.completedGeneration` are updated directly without launching the
+  worker `job`.
+
+The **migrator worker** `job` runs the equivalence of `oc adm migrate storage
+--include=<resource type>` to migrate a resource type. The **migrator worker**
+uses API chunking to retrieve partial lists of a resource type and thus can
+migrate a small chunk at a time. It stores the [continue token] in a `configmap`
+to preserve progress. With the inconsistent continue token introduced in
+[#67284][], the **migrator worker** does not need to worry about expired continue
+token.
+
+[#67284]:https://github.com/kubernetes/kubernetes/pull/67284
+[API section]:#api
+
+The cluster admin can run the `kube-migrator progress` command to check if the
+migration has completed. The command lists all `migrations` instances and checks
+if all of them have `.spec.generation == .status.observedGeneration`.
+
+### API
+
+We introduce the `migration` API to record the intention and progress of a
+migration. The API will be a CRD defined in the `migration.k8s.io` group.
+
+Read the [workflow section][] to understand how the API is used.
+
+```golang
+type StorageVersionMigration struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+  Spec StorageVersionMigrationSpec
+  Status StorageVersionMigrationStatus
+}
+
+// Note that the apiserver will increment `metadata.generation` when this spec
+// is updated.
+type StorageVersionMigrationSpec {
+  // Resource is the identifier of the resource that's being migrated.
+  // TBD: maybe resource=hash(etcd instance, etcd path prefix).
+  // Immuntable.
+  Resource string 
+
+  // GroupVersions are all the groups and versions the apiserver serves this
+  // resource in. The server preferred version should be groupVersions[0].
+  // **Migration worker** GET & PUT via the preferred version by default, to
+  // avoid conversions in the apiserver.
+  GroupVersions []GroupVersion
+  
+  // LastUpdateTimestamp is the last time this spec is updated. The value is set
+  // by clients, not managed by the apiserver. The timestamp introduces randomness
+  // to the spec so that every update of the spec changes the contents of the spec,
+  // which leads to the increment of `metadata.generation`.
+  LastUpdateTimestamp metav1.Time
+
+  // Reason of the migration. Examples are "manually triggered", "apiserver
+  // updated", etc.
+  Reason string
+}
```

Comment:
I guess what I expected to see is an object that expresses a migration request for a single resource type (e.g., apps/v1/deployment). With status indicating progress {not started, migrated N objects, completed}, and where the object is removed after completion.

The generation stuff makes me think you want to reuse a single object to make multiple requests, but I don't understand why you would want to do that.